From a9e0bdc021511c77397c5c2730ec2c9e452a6486 Mon Sep 17 00:00:00 2001
From: Matteo Melli <matteom@ongres.com>
Date: Sun, 12 Apr 2026 14:52:05 +0000
Subject: [PATCH] patch: add a build option to disable contrib builds

Per the Postgres docs, `make world` does build contrb but `make all`
skips it. However, the meson build does build contrib by default.
---
 meson.build       | 19 +++++++++++++++++--
 meson_options.txt |  3 +++
 2 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/meson.build b/meson.build
index 366b347094..1efec656c0 100644
--- a/meson.build
+++ b/meson.build
@@ -3378,7 +3378,20 @@ test_install_libs = []
 
 subdir('src')
 
-subdir('contrib')
+if get_option('contrib')
+  subdir('contrib')
+else
+  # NOTE:
+  # These variables are normally provided by contrib meson.build files but are
+  # used elsewhere in the build, so they must be redefined here when contrib is
+  # disabled.
+
+  # defined in contrib/meson.build, used in install_test_files_args
+  contrib_data_dir = dir_data_extension
+
+  # defined in contrib/spi/meson.build, used in src/test/regress/meson.build
+  refint_cflags = ['-DREFINT_VERBOSE']
+endif
 
 subdir('src/test')
 subdir('src/interfaces/ecpg/test')
@@ -3838,7 +3851,9 @@ endif
 alias_target('backend', backend_targets)
 alias_target('bin', bin_targets + [libpq_st])
 alias_target('pl', pl_targets)
-alias_target('contrib', contrib_targets)
+if get_option('contrib')
+  alias_target('contrib', contrib_targets)
+endif
 alias_target('testprep', testprep_targets)
 
 alias_target('world', all_built, docs)
diff --git a/meson_options.txt b/meson_options.txt
index 9e3be4a8ab..3d132a3834 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -73,6 +73,9 @@ option('rpath', type: 'boolean', value: true,
 option('prefix_distro', type: 'string', value: '',
   description: 'Select a "distro prefix" (a path different from prefix where Postgres will be installed). Useful if the distro install path will be different from the install path at build time.')
 
+option('contrib', type: 'boolean', value: true,
+  description: 'Build contrib modules')
+
 # External dependencies
 
 option('bonjour', type: 'feature', value: 'auto',
